File: Algorithms/p202_Happy_Number.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        table = [False]*811
        while True:
            m = 0
            while n:
                m += (n%10)**2
                n //= 10
            n = m
            if n==1:
                return True
            if table[n]:
                return False
            else:
                table[n] = True
        return False # dummy

File: Algorithms/p202_Happy_Number.cpp
class Solution {
public:
    bool isHappy(int n) {
        bool table[810] = {};
        while(1){
            int m = 0;
            while(n){
                int d = n%10;
                m += d*d;
                n /= 10;
            }
            n = m;
            if(n==1){
                return 1;
            }
            if(table[n]){
                return 0;
            }else{
                table[n] = 1;
            }
        }
        return 0; // dummy
    }
};
